### PR TITLE
[solver-interface 2/n]: Minion Implementation

### LINKED ISSUE #188: solver-interface: Minion Implementation

### PR DESCRIPTION
**Pull request by niklasdewally**
This PR continues work on the new solver interface (#168, #176) by adding a Minion adaptor implementation.

I have also added a documentation example, and reworked the types a bit so that solvers receive callbacks as thread-safe closures (thank you for the feedback @gskorokhod).

Closes #188.

**Testing**

* The integration tester is now using the new interface successfully.
* The module-level documentation example is runnable as a doc-test. This counts the number of solutions in the bool-03 integration test  (thank you again @PedroGGBM for the wonderful get_example function!).

**Reviewing this PR**

A large part of this PR is documentation, example code, and interface design. Therefore, it would be helpful to me if during reviewing you run `cargo doc`, read the module level documentation ,and let me know if anything seems unclear in the example code or prose.

### CODE REVIEW

**Comment by niklasdewally** [comment]
Not sure who to mark as reviewer for this one; would appreciate reviews from anyone who has time :) @PedroGGBM @gskorokhod @Kieranoski702 @lixitrixi

File: conjure_oxide/src/unstable/solver_interface/adaptors/minion.rs [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+    fn solve(
+        &mut self,
+        model: Self::Model,
+        callback: SolverCallback,
+        _: private::Internal,
+    ) -> Result<SolveSuccess, SolverError> {
+        // our minion callback is global state, so single threading the adaptor as a whole is
+        // probably a good move...
+        #[allow(clippy::unwrap_used)]
+        let mut locked = MINION_LOCKED.lock().unwrap();
```
Comment (ChrisJefferson) [inline conjure_oxide/src/unstable/solver_interface/adaptors/minion.rs:68]: Why can't the locking just be a Mutex<()>, you lock it here, keep it until the end of this function, and when you drop it Rust will automatically give it to someone else (so no need for the notify?)

File: conjure_oxide/src/unstable/solver_interface/adaptors/minion.rs [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+            minion_model
+                .constraints
+                .push(minionast::Constraint::WatchNeq(
+                    read_var(*a)?,
+                    read_var(*b)?,
+                ));
+            Ok(())
+        }
+        x => Err(ModelFeatureNotSupported(format!("{:?}", x))),
+    }
```
Comment (ChrisJefferson) [inline conjure_oxide/src/unstable/solver_interface/adaptors/minion.rs:255]: This is fine for now, but long term this should probably be autogenerated, or done dynamically (minion generates this as a big JSON file. It would be easy to make Minion output that JSON file, I don't think it's currently stored somewhere nciely.

File: conjure_oxide/src/unstable/solver_interface/states.rs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+/// The state returned by [`Solver`] if solving has been successful.
+pub struct ExecutionSuccess {
+    /// Execution statistics.
+    pub stats: Option<Box<dyn Stats>>,
+
+    /// Cannot construct this from outside this module.
+    pub _sealed: Internal,
+}
+
+/// The state returned by [`Solver`] if solving has not been successful.
```
Comment (ChrisJefferson) [inline conjure_oxide/src/unstable/solver_interface/states.rs:36]: What does "not successful" mean? Is it some kind of solver error, or there just weren't any solutions? If it's "no solutions", did we just run out of time (if there was a time limit)? I'd be clearer which is going on

File: conjure_oxide/src/unstable/solver_interface/adaptors/minion.rs (hunk 1 above)
Comment (niklasdewally) [reply conjure_oxide/src/unstable/solver_interface/adaptors/minion.rs:68]: I've never been fully sure about the difference between convar and mutexes - I agree that a mutex would probably be good enough here.

I guess the performance impact of condvars.comes into play when we want to lock depending on the value inside the mutex? This would avoid busy waits like while (a.lock() == false) {}? I don't need a value, just mutual exclusion, so this is irrelevant...

Thank you for clarifying this for me :)

File: conjure_oxide/src/unstable/solver_interface/states.rs (hunk 3 above)
Comment (niklasdewally) [reply conjure_oxide/src/unstable/solver_interface/states.rs:36]: Maybe I could have type states for SolverFinish and SolverCrash, and then we can subdivide SolverFinish into an arbitrary number of  subclasses such as sat, unsat, timeout, ...

Comment (ChrisJefferson) [reply conjure_oxide/src/unstable/solver_interface/states.rs:36]: I'd keep solver errors seperate, because I suspect most people will want to `unwrap` those (because solver error generally means something horrible has happened, and it's probably best to just abort the program altogether).

With regards timeout vs solutions, there isn't a perfect answer. I'd be tempted to just always return a list of solutions (which might be empty, or have 1, or have many, depending on what was asked for). While a list of length 1 might be annoying for people who only want one solution, it's also a bit annoying to split 1 solutions vs many (unless there is going to be different search methods to start a search for 1 solution vs many?) and then also have SolverFinished split into complete or incomplete (incomplete can be because of timeout, or some kind of "search/node limit").

File: conjure_oxide/src/unstable/solver_interface/adaptors/minion.rs (hunk 1 above)
Comment (ChrisJefferson) [reply conjure_oxide/src/unstable/solver_interface/adaptors/minion.rs:68]: Generally, a convar is for when you need to wakeup / signal another thread (for example, there is some work todo), here you just need mutual exclusion, so a Mutex should be fine by itself (you could actually put some data in the mutex, usually in Rust we do, but something you find yourself protecting an external resource, and then you just use Mutex<()>).

File: conjure_oxide/src/unstable/solver_interface/states.rs (hunk 3 above)
Comment (niklasdewally) [reply conjure_oxide/src/unstable/solver_interface/states.rs:36]: I agree with all this, thank you for the feedback :)

We are getting solutions out by callback, and give the user custom logic to handle/ not handle them / abort search early - therefore I don't think we can have a list of them in a clean way, and the user should do this themselves if they want it.
 
I think we decided not to give a list for memory concerns as a model can have a large number of solutions - a good user coded callback could just drop solutions deemed to be identical or useless from the list, but we can't do that in generality (at this level of the program).


However I can definitely keep track of if any solutions have been found, and say "you got solutions" vs "you got no solutions". 

Incomplete and complete is definitely a  good taxonomy - would "user aborted the search" be counted as complete?

File: conjure_oxide/tests/integration/xyz/input.expected-minion.solutions.json [hunk 4]
```diff
@@ -1,12 +1,12 @@
...
-    "c": 2
+    "UserName(a)": 1,
+    "UserName(b)": 1,
+    "UserName(c)": 2
   },
   {
-    "a": 2,
-    "b": 1,
-    "c": 1
+    "UserName(a)": 2,
```
Comment (ozgurakgun) [inline conjure_oxide/tests/integration/xyz/input.expected-minion.solutions.json:8]: is this change on purpose? no problem if so.

Comment (niklasdewally) [reply conjure_oxide/tests/integration/xyz/input.expected-minion.solutions.json:8]: Minion / minion_rs stores names as strings always, but conjure AST has different types of names (Username, MachineName). 

Previously, the minion_rs AST types were used to get results, and the tester JSON used them for its output as well.
Now the tester  and its generated output uses conjure AST types. This means that the outputted names can be MachineNames or UserNames.


(as discussed in meeting)

**Comment by niklasdewally** [comment]
Todo before merge:
1. Store wall time in the solver state after execution. (CPU time is tricky to do in generality, so we decided to make this optional and part of the solver specific stats)

2. New taxonomy of exit status as suggested by Chris - currently on my local machine.

3. Add in changes and additions from Felix's PR when merged, including machine names.

4. Documentation on how new solver adaptors should be implemented, and what rules they should follow (eg they must keep track of if any solutions were found or not and use the appropriate output state).

**Comment by niklasdewally** [comment]
I'm waiting for #230 to be merged (so that Felix doesn't have to deal with even more merge conflicts!); but the rest is done now
